File: lua/tst/in.lua
pico.init(true)
pico.set.scene { dim = {'!', w=100, h=100} }

print "pico.xin.rect smoke"
do
    local out   = {'!', x=20, y=20, w=60, h=60, anchor='NW'}
    local child = {'%', x=0.5, y=0.5, w=0.5, h=0.5, anchor='C'}
    local ret = pico.xin.rect(out, child)
    assert(ret[1] == '%')
    assert(ret.anchor.x == 0.5 and ret.anchor.y == 0.5)
    local abs = pico.cv.rect(nil, '!', nil, ret)
    assert(math.floor(abs.x+0.5)==50 and math.floor(abs.y+0.5)==50)
    assert(math.floor(abs.w+0.5)==30 and math.floor(abs.h+0.5)==30)
end

print "pico.xin.pos smoke"
do
    local out   = {'!', x=20, y=20, w=60, h=60}
    local child = {'%', x=0.5, y=0.5}
    local ret = pico.xin.pos(out, child)
    assert(ret[1] == '%')
    local abs = pico.cv.pos(nil, '!', nil, ret)
    assert(abs.x==20.5 and abs.y==20.5)
end

print "pico.xin.dim smoke"
do
    local out   = {'!', x=20, y=20, w=60, h=60, anchor='NW'}
    local child = {'%', w=0.5, h=0.5}
    local ret = pico.xin.dim(out, child)
    assert(ret[1] == '%')
    local abs = pico.cv.dim(nil, '!', nil, ret)
    assert(math.floor(abs.w+0.5)==30 and math.floor(abs.h+0.5)==30)
end

print "\n=== ALL TESTS PASSED ==="
pico.init(false)

File: tst/in.c
#include "pico.h"
#include <stdio.h>
#include <assert.h>

int main (void) {
    pico_init(1);
    Pico_Rel_Dim log = { '!', {100, 100} };
    pico_set_scene_dim(&log);

    // RECT IN RECT - % C inside ! NW (composition)
    {
        puts("rect in rect - % C in ! NW");
        Pico_Rel_Rect out = { '!', {20, 20, 60, 60}, PICO_ANCHOR_NW };
        Pico_Rel_Rect in  = { '%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C };
        Pico_Rel_Rect ret = pico_in_rect(out, in);
        Pico_Rel_Rect abs = { '!', {0, 0, 0, 0}, PICO_ANCHOR_NW };
        pico_cv_rect(NULL, &abs, NULL, &ret);
        assert((int)(abs.x+0.5f)==35 && (int)(abs.y+0.5f)==35);
        assert((int)(abs.w+0.5f)==30 && (int)(abs.h+0.5f)==30);
    }

    // POS IN RECT - % NW
    {
        puts("pos in rect - % NW");
        Pico_Rel_Rect out = { '!', {20, 20, 60, 60}, PICO_ANCHOR_NW };
        Pico_Rel_Pos  in  = { '%', {0.5, 0.5}, PICO_ANCHOR_NW };
        Pico_Rel_Pos  ret = pico_in_pos(out, in);
        Pico_Rel_Pos abs = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos(NULL, &abs, NULL, &ret);
        assert((int)(abs.x+0.5f)==50 && (int)(abs.y+0.5f)==50);
    }

    // DIM IN RECT - %
    {
        puts("dim in rect - %");
        Pico_Rel_Rect out = { '!', {20, 20, 60, 60}, PICO_ANCHOR_NW };
        Pico_Rel_Dim  in  = { '%', {0.5, 0.5} };
        Pico_Rel_Dim  ret = pico_in_dim(out, in);
        Pico_Rel_Dim abs = { '!', {0, 0} };
        pico_cv_dim(NULL, &abs, NULL, &ret);
        assert((int)(abs.w+0.5f)==30 && (int)(abs.h+0.5f)==30);
    }

    // MODE / ANCHOR PRESERVED FROM in
    {
        puts("mode and anchor preserved");
        Pico_Rel_Rect out = { '!', {20, 20, 60, 60}, PICO_ANCHOR_NW };
        Pico_Rel_Rect in  = { '%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C };
        Pico_Rel_Rect ret = pico_in_rect(out, in);
        assert(ret.mode == in.mode);
        assert(ret.anchor.x == in.anchor.x);
        assert(ret.anchor.y == in.anchor.y);
    }

    printf("\n=== ALL TESTS PASSED ===\n");
    pico_init(0);
    return 0;
}
